src/code/z_eff_shield_particle.c

```diff
@@ -1,11 +1,211 @@
+
+    if ((this != NULL) && (this->lightDecay == true)) {
+        if (this->lightNode == Effect_GetGlobalCtx()->lightCtx.listHead) {
+            Effect_GetGlobalCtx()->lightCtx.listHead = this->lightNode->next;
+        }
+        LightContext_RemoveLight(Effect_GetGlobalCtx(), &Effect_GetGlobalCtx()->lightCtx, this->lightNode);
+    }
+}
+
+s32 EffectShieldParticle_Update(void* thisx) {
+    EffectShieldParticle* this = (EffectShieldParticle*)thisx;
+    EffectShieldParticleElement* elem;
+
+    if (this == NULL) {
+        return 0;
+    }
+
+    for (elem = &this->elements[0]; elem < &this->elements[this->numElements]; elem++) {
+        elem->endXChange -= this->deceleration;
+        if (elem->endXChange < 0.0f) {
+            elem->endXChange = 0.0f;
+        }
+
+        if (elem->startXChange > 0.0f) {
+            elem->startXChange -= this->deceleration;
+            if (elem->startXChange < 0.0f) {
+                elem->startXChange = 0.0f;
+            }
+        }
+
+        elem->endX += elem->endXChange;
+        elem->startX += elem->startXChange;
+
+        if ((elem->startXChange == 0.0f) && (this->lengthCutoff < (elem->endX - elem->startX))) {
+            elem->startXChange = elem->initialSpeed;
+        }
+    }
+
+    if (this->lightDecay == true) {
+        this->lightInfo.params.point.radius /= 2;
+    }
+
+    this->timer++;
+
+    if (this->duration < this->timer) {
+        return 1;
+    }
+
+    return 0;
+}
+
+void EffectShieldParticle_GetColors(EffectShieldParticle* this, Color_RGBA8* primColor, Color_RGBA8* envColor) {
+    s32 halfDuration = this->duration * 0.5f;
+    f32 ratio;
+
+    if (halfDuration == 0) {
+        primColor->r = this->primColorStart.r;
+        primColor->g = this->primColorStart.g;
+        primColor->b = this->primColorStart.b;
+        primColor->a = this->primColorStart.a;
+        envColor->r = this->envColorStart.r;
+        envColor->g = this->envColorStart.g;
+        envColor->b = this->envColorStart.b;
+        envColor->a = this->envColorStart.a;
+    } else if (this->timer < halfDuration) {
+        f32 ratio = this->timer / (f32)halfDuration;
+        primColor->r = this->primColorStart.r + (this->primColorMid.r - this->primColorStart.r) * ratio;
+        primColor->g = this->primColorStart.g + (this->primColorMid.g - this->primColorStart.g) * ratio;
+        primColor->b = this->primColorStart.b + (this->primColorMid.b - this->primColorStart.b) * ratio;
+        primColor->a = this->primColorStart.a + (this->primColorMid.a - this->primColorStart.a) * ratio;
+        envColor->r = this->envColorStart.r + (this->envColorMid.r - this->envColorStart.r) * ratio;
+        envColor->g = this->envColorStart.g + (this->envColorMid.g - this->envColorStart.g) * ratio;
+        envColor->b = this->envColorStart.b + (this->envColorMid.b - this->envColorStart.b) * ratio;
+        envColor->a = this->envColorStart.a + (this->envColorMid.a - this->envColorStart.a) * ratio;
```

Comment:
```suggestion
        primColor->r = F32_LERPIMP(this->primColorStart.r, this->primColorMid.r, ratio);
        primColor->g = F32_LERPIMP(this->primColorStart.g, this->primColorMid.g, ratio);
        primColor->b = F32_LERPIMP(this->primColorStart.b, this->primColorMid.b, ratio);
        primColor->a = F32_LERPIMP(this->primColorStart.a, this->primColorMid.a, ratio);
        envColor->r = F32_LERPIMP(this->envColorStart.r, this->envColorMid.r, ratio);
        envColor->g = F32_LERPIMP(this->envColorStart.g, this->envColorMid.g, ratio);
        envColor->b = F32_LERPIMP(this->envColorStart.b, this->envColorMid.b, ratio);
        envColor->a = F32_LERPIMP(this->envColorStart.a, this->envColorMid.a, ratio);
```

Reply:
I made a new `LERPIMP` which isn't restricted to floats like you suggested, but I also left the `F32_LERPIMP` for float purposes and added f32 casts to it so it forces floats to be used and avoids casts in the case of `z_eff_spark.c`